File: ios/Sources/PoltioSDK/PoltioTriggerDismissalStore.swift
import Foundation

/// Persists explicit user dismissals of the pill/box "remember for N hours" close button
/// (`pillCloseRememberDuration` / `boxCloseRememberDuration`), so a widget the user explicitly closed
/// doesn't reappear until the remember window elapses. Mirrors the `UserDefaults`-backed persistence
/// pattern already used for `sdkId`/`puid` in `PoltioSDK`.
enum PoltioTriggerDismissalStore {
    private static let storageKey = "com.poltio.sdk.trigger_dismissals"

    /// Returns whether `publicId` is currently within an active "remember" dismissal window.
    static func isDismissed(publicId: String, now: Date = Date()) -> Bool {
        guard let dismissedUntil = load()[publicId] else {
            return false
        }
        return now.timeIntervalSince1970 < dismissedUntil
    }

    /// Records that `publicId` was explicitly closed; it will be suppressed for `hours` from now.
    static func recordDismissal(publicId: String, hours: Double, now: Date = Date()) {
        guard hours > 0 else { return }
        var stored = load()
        stored[publicId] = now.timeIntervalSince1970 + hours * 3600
        save(stored)
    }

    #if DEBUG
        /// Clears all persisted dismissals. Used exclusively for unit testing.
        static func reset() {
            UserDefaults.standard.removeObject(forKey: storageKey)
        }
    #endif

    private static func load() -> [String: TimeInterval] {
        UserDefaults.standard.dictionary(forKey: storageKey) as? [String: TimeInterval] ?? [:]
    }

    private static func save(_ dict: [String: TimeInterval]) {
        UserDefaults.standard.set(dict, forKey: storageKey)
    }
}

File: android/poltio-sdk/src/main/java/com/poltio/sdk/PoltioTriggerDismissalStore.kt
package com.poltio.sdk

import android.content.Context
import org.json.JSONObject

/**
 * Persists explicit user dismissals of the pill/box "remember for N hours" close button
 * (`pillCloseRememberDuration` / `boxCloseRememberDuration`), so a widget the user explicitly
 * closed doesn't reappear until the remember window elapses. Mirrors the `SharedPreferences`-backed
 * persistence pattern already used for `sdkId`/`puid` in [PoltioSDK].
 */
internal object PoltioTriggerDismissalStore {
    private const val PREFS_NAME = "com.poltio.sdk.prefs"
    private const val STORAGE_KEY = "trigger_dismissals"

    /** Returns whether [publicId] is currently within an active "remember" dismissal window. */
    fun isDismissed(context: Context, publicId: String, nowMs: Long = System.currentTimeMillis()): Boolean {
        val dismissedUntil = load(context)[publicId] ?: return false
        return nowMs < dismissedUntil
    }

    /** Records that [publicId] was explicitly closed; it will be suppressed for [hours] from now. */
    fun recordDismissal(context: Context, publicId: String, hours: Double, nowMs: Long = System.currentTimeMillis()) {
        if (hours <= 0) return
        val stored = load(context).toMutableMap()
        stored[publicId] = nowMs + (hours * 3_600_000.0).toLong()
        save(context, stored)
    }

    /** Clears all persisted dismissals. Used exclusively for unit testing. */
    fun reset(context: Context) {
        prefs(context).edit().remove(STORAGE_KEY).apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private fun load(context: Context): Map<String, Long> {
        val json = prefs(context).getString(STORAGE_KEY, null) ?: return emptyMap()
        return try {
            val obj = JSONObject(json)
            val result = LinkedHashMap<String, Long>()
            obj.keys().forEach { key -> result[key] = obj.getLong(key) }
            result
        } catch (error: Exception) {
            emptyMap()
        }
    }

    private fun save(context: Context, map: Map<String, Long>) {
        val obj = JSONObject()
        map.forEach { (key, value) -> obj.put(key, value) }
        prefs(context).edit().putString(STORAGE_KEY, obj.toString()).apply()
    }
}
